src/types/function.rs

```diff
@@ -173,10 +117,64 @@ impl PyCFunction {
         } else {
             (std::ptr::null_mut(), std::ptr::null_mut())
         };
-        Self::internal_new_from_pointers(method_def, py, mod_ptr, module_name)
+        let (def, destructor) = method_def.as_method_def()?;
+
+        // FIXME: stop leaking the def and destructor
+        let def = Box::into_raw(Box::new(def));
+        std::mem::forget(destructor);
+
+        unsafe {
+            py.from_owned_ptr_or_err::<PyCFunction>(ffi::PyCFunction_NewEx(
+                def,
+                mod_ptr,
+                module_name,
+            ))
+        }
     }
 }
 
+fn closure_capsule_name(py: Python<'_>) -> &'static CStr {
+    static CLOSURE_CAPSULE_NAME: GILOnceCell<&CStr> = GILOnceCell::new();
+    CLOSURE_CAPSULE_NAME.get_or_init(py, || CStr::from_bytes_with_nul(b"pyo3-closure\0").unwrap())
```

Comment:
Does it really help to memoize calling `CStr::from_bytes_with_nul` using a byte slice literal? I would almost expect the compiler to fully evaluate this at compile time.

Reply:
Good point, I'll reduce this to just call `CStr::from_bytes_with_nul(b"pyo3-closure\0").unwrap()` and then merge.